include/zephyr/bluetooth/gatt.h

```diff
@@ -167,16 +167,57 @@ typedef ssize_t (*bt_gatt_attr_write_func_t)(struct bt_conn *conn,
 
 /** @brief GATT Attribute structure. */
 struct bt_gatt_attr {
-	/** Attribute UUID */
+	/** @brief Attribute Type, aka. "UUID" */
 	const struct bt_uuid *uuid;
-	/** Attribute read callback */
+
+	/** @brief Attribute Value function
+	 *
+	 * This is the general API to read from the Attribute Value of
+	 * this Attribute in on-air format.
+	 *
+	 * This function may safely assume the Attribute Permissions are
+	 * satisfied for this read.
+	 *
+	 * This function is primarily invoked by the GATT server when a
+	 * remote GATT client performs a read, but it can also be
+	 * invoked for a local read.
+	 *
+	 * May be NULL if the attribute is not readable.
+	 */
 	bt_gatt_attr_read_func_t read;
-	/** Attribute write callback */
+
+	/** @brief Attribute Value write method
+	 *
+	 * This is the general API to write to the Attribute Value of
+	 * this Attribute in on-air format.
+	 *
+	 * This function may safely assume the Attribute Permissions are
+	 * satisfied for this write.
+	 *
+	 * May be NULL if the attribute is not writable.
+	 */
 	bt_gatt_attr_write_func_t write;
-	/** Attribute user data */
+
+	/** @brief Userdata for the read() and write()
+	 *
+	 * The meaning of this field is not specified here.
+	 *
+	 * @warning Do not assume the meaning of this field based just
+	 * on the Attribute Type. Two Attributes may share Attribute
+	 * Type but have different internal implementations.
+	 *
+	 * @note Attributes in callbacks from bt_gatt_discover() have
+	 * speficied meanings for this field. See @ref
+	 * bt_gatt_discover_func_t.
+	 */
 	void *user_data;
-	/** Attribute handle */
+
+	/** @brief Attribute Handle (unless zero)
+	 *
+	 * To get the handle, use bt_gatt_attr_get_handle() instead.
```

Comment:
Could you elaborate on why we should call `bt_gatt_attr_get_handle()` instead of reading it here?

Reply:
Mainly because this field does not always contain the handle. It's always zero (the spec reserved handle, which we abuse to mean "not a handle") for attributes in statically defined GATT services.

But it's enough to just say that `bt_gatt_attr_get_handle` is the getter function, and the field shall not be read directly.

Unfortunately, this field is extra complicated, because if it's not zero at registration time, it explicitly determines the handle instead of the stack allocating one. This feature is obscure and causes confusion: https://github.com/zephyrproject-rtos/zephyr/issues/67377

Should we include any of this info in the docs?

Reply:
I see. Maybe we should deprecate this field and make internal instead? (prefix with underscore)
Another option is to document when it is safe to access the field. But I guess that is complex